cmd/zoekt-sourcegraph-indexserver/main.go

```
@@ -322,6 +330,58 @@ func (s *Server) Run(queue *Queue) {
 	}
 }
 
+// Update priority.json given new entries, and remove no longer tracked repos.
+// This doesn't simply write newPriorities because a transient getIndexOptions failure
+// would cause the associated repo to get deprioritized.
+func (s *Server) maybeUpdatePriorities(names []string, newPriorities map[string]float64) {
+	priorityPath := path.Join(s.IndexDir, "priority.json")
+	priorities := map[string]float64{}
+	buf, err := ioutil.ReadFile(priorityPath)
+	if err == nil {
+		err = json.Unmarshal(buf, &priorities)
+		if err != nil {
+			log.Printf("error loading old priority.json: %v", err)
+		}
+	}
+
+	// maybe remove no-longer-tracked repos from the priorities list
+	if len(names) != len(priorities) {
+		set := make(map[string]struct{}, len(names))
+		for _, name := range names {
+			set[name] = struct{}{}
+		}
+		for name := range priorities {
+			if _, ok := set[name]; !ok {
+				delete(priorities, name)
+			}
+		}
+	}
+
+	for name, priority := range newPriorities {
+		priorities[name] = priority
+	}
+
+	newBuf, err := json.Marshal(priorities)
+	if err != nil {
+		log.Printf("error marshaling new priority.json: %v", err)
+	}
+	newBuf = append(newBuf, '\n') // prettier
+
+	if bytes.Equal(buf, newBuf) {
+		return // no need to rewrite priority.json
+	}
+
+	err = ioutil.WriteFile(priorityPath+".tmp", newBuf, 0644)
+	if err != nil {
+		log.Printf("error writing new priority.json: %v", err)
```

Should we return an error instead and log it in the returning function?